server/src/main/java/com/google/musicanalysis/site/OAuthLoginServlet.java

```diff
@@ -12,19 +15,35 @@
 
 @WebServlet("/api/oauth/login/*")
 public class OAuthLoginServlet extends HttpServlet {
+  private static final Logger LOGGER = Logger.getLogger(OAuthLoginServlet.class.getName());
 
   @Override
   protected void doGet(HttpServletRequest req, HttpServletResponse res)
       throws ServletException, IOException {
-    var path = req.getPathInfo().substring(1);
+    var path = req.getPathInfo();
 
     if (path == null) {
       res.setStatus(404);
       return;
     }
 
     // URI that user should be redirected to after logging in
-    var redirectUri = System.getenv().get("OAUTH_CALLBACK_URI");
+    URI domainUri;
+
+    try {
+      domainUri = new URI(System.getenv().get("DOMAIN"));
+    } catch (URISyntaxException e) {
+      LOGGER.severe("The DOMAIN environment variable is invalid.");
+      res.setStatus(500);
+      return;
+    } catch (NullPointerException e) {
+      LOGGER.severe("The DOMAIN environment variable is not set.");
+      res.setStatus(500);
+      return;
+    }
+
+    var redirectUri = domainUri.resolve("/api/oauth/callback");
```

Comment:
Used this link to look up `URI.resolve()` -> https://stackoverflow.com/questions/2534124/java-uri-resolve

Why is this necessary in this case? Asking for knowledge.

Reply:
I always like to use methods that have special handling when I'm concatenating paths. In this case, I could just do a string concatenation and probably get away with it: "localhost:8080" + "/api/oauth/callback" = "localhost:8080/api/oauth/callback". But what if for some reason the first part is specified as "localhost:8080/"? That's technically the same path, but doing a string concatenation would give "localhost:8080//api/oauth/callback", which is not a valid path. `URI.resolve()` will handle this case automatically.